x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/CategorizeBlockHashTests.java

```diff
@@ -95,41 +95,114 @@ public void testCategorizeRaw() {
-                @Override
-                public void add(int positionOffset, IntBlock groupIds) {
-                    assertEquals(groupIds.getPositionCount(), positions);
+        try (var hash = new CategorizeBlockHash(blockFactory, 0, AggregatorMode.SINGLE, analysisRegistry)) {
+            for (int i = randomInt(2); i < 3; i++) {
+                hash.add(page, new GroupingAggregatorFunction.AddInput() {
+                    @Override
+                    public void add(int positionOffset, IntBlock groupIds) {
+                        assertEquals(groupIds.getPositionCount(), positions);
+
+                        assertEquals(1, groupIds.getInt(0));
+                        assertEquals(2, groupIds.getInt(1));
+                        assertEquals(2, groupIds.getInt(2));
+                        assertEquals(2, groupIds.getInt(3));
+                        assertEquals(3, groupIds.getInt(4));
+                        assertEquals(1, groupIds.getInt(5));
+                        assertEquals(1, groupIds.getInt(6));
+                        if (withNull) {
+                            assertEquals(0, groupIds.getInt(7));
+                        }
+                    }
 
-                    assertEquals(1, groupIds.getInt(0));
-                    assertEquals(2, groupIds.getInt(1));
-                    assertEquals(2, groupIds.getInt(2));
-                    assertEquals(2, groupIds.getInt(3));
-                    assertEquals(3, groupIds.getInt(4));
-                    assertEquals(1, groupIds.getInt(5));
-                    assertEquals(1, groupIds.getInt(6));
-                    if (withNull) {
-                        assertEquals(0, groupIds.getInt(7));
+                    @Override
+                    public void add(int positionOffset, IntVector groupIds) {
+                        add(positionOffset, groupIds.asBlock());
                     }
-                }
 
-                @Override
-                public void add(int positionOffset, IntVector groupIds) {
-                    add(positionOffset, groupIds.asBlock());
-                }
+                    @Override
+                    public void close() {
+                        fail("hashes should not close AddInput");
+                    }
+                });
 
-                @Override
-                public void close() {
-                    fail("hashes should not close AddInput");
-                }
-            });
+                assertHashState(hash, withNull, ".*?Connected.+?to.*?", ".*?Connection.+?error.*?", ".*?Disconnected.*?");
+            }
         } finally {
             page.releaseBlocks();
         }
 
-        // TODO: randomize and try multiple pages.
-        // TODO: assert the state of the BlockHash after adding pages. Including the categorizer state.
-        // TODO: also test the lookup method and other stuff.
+        // TODO: randomize values? May give wrong results
```

Comment:
I could see randomizing the values in a separate test and asserting that we get expected results/don't crash out.

Reply:
But later. This is fine now.